File: src/main/kotlin/interview_150/array/MajorityElement.java
package interview_150.array;


import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Difficulty : Medium
 * Problem Statement : https://leetcode.com/problems/majority-element/description/?envType=study-plan-v2&envId=top-interview-150
 */

public class MajorityElement {
    public int majorityElement(int[] nums) {
        int size = nums.length;

        Map<Integer, Integer> map = new HashMap<Integer, Integer>();

        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        size = size / 2;

        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() > size) {
                return entry.getKey();
            }
        }
        return 0;
    }

    public int majorityElement2(int[] nums) {
        Arrays.sort(nums);

        int n = nums.length;
        return nums[n / 2];
    }
}

File: src/main/kotlin/two_weeks_study_plan/data_structure/data_structure_2/day_1/MajorityElement.kt
package two_weeks_study_plan.data_structure.data_structure_2.day_1

/**
 * Difficulty : Easy
 * Problem Statement : https://leetcode.com/problems/majority-element/?envType=study-plan&id=data-structure-ii
 */

fun main() {
    println(MajorityElement().majorityElement2(intArrayOf(3, 2, 3)))
}

class MajorityElement {
    fun majorityElement(nums: IntArray): Int {
        val frequencyMap = nums.groupBy { it }.map { (key, value) -> key to value.size }.toMap()

        return frequencyMap.maxByOrNull { it.value }!!.key
    }

    fun majorityElement2(nums: IntArray): Int {
        var count = 0
        var candidate: Int? = null

        nums.forEach {
            if (count == 0) candidate = it

            count += if (candidate == it) 1 else -1
        }
        return candidate!!
    }

}